Source/NSubstitute/Core/CallSpecificationFactory.cs

```diff
@@ -15,7 +15,7 @@ public ICallSpecification CreateFrom(ICall call, MatchArgs matchArgs)
         {
             var methodInfo = call.GetMethodInfo();
             var argumentSpecs = call.GetArgumentSpecifications();
-            var arguments = call.GetArguments();
+            var arguments = call.GetOriginalArguments();
```

Comment:
@alexandrnikitin @dtchepak After I investigated that more, I found that already applied changes are not enough to fully cover this sub-feature. Thank you for pointing me to this direction. If we use original arguments for specification composition, we should also ensure that we take original arguments for matching the calls against the specification. 

Therefore, I was forced to add [this commit](https://github.com/nsubstitute/NSubstitute/pull/259/commits/26994b8ab12426001073dd62c2c8ff8b972c4b9f). Please see the changed tests - all them are failing without the introduced changes.

Sorry Alexandr, now you will need to thing about even more set of changes 😉  But, we definitely need. Current behavior of NSubstitute is weak in that area.

P.S. I like how this PR evolves and becomes more mature and consistent.